src/server/scripts/Kalimdor/RuinsOfAhnQiraj/boss_moam.cpp

```diff
@@ -16,178 +16,282 @@
+enum MoamSpells
 {
+    // Combat
     SPELL_TRAMPLE               = 15550,
-    SPELL_DRAIN_MANA            = 25671,
+    SPELL_DRAIN_MANA            = 25676,
     SPELL_ARCANE_ERUPTION       = 25672,
-    SPELL_SUMMON_MANA_FIEND_1   = 25681, // TARGET_DEST_CASTER_FRONT
-    SPELL_SUMMON_MANA_FIEND_2   = 25682, // TARGET_DEST_CASTER_LEFT
-    SPELL_SUMMON_MANA_FIEND_3   = 25683, // TARGET_DEST_CASTER_RIGHT
-    SPELL_ENERGIZE              = 25685
+    SPELL_SUMMON_MANA_FIENDS    = 25684,
+    SPELL_ENERGIZE              = 25685,
+
+    // Misc
+    SPELL_ZERO_MANA_FULL_HEALTH = 23777,
+    SPELL_DROP_OBSIDIAN         = 27630,
+
+    // Scripts
+    SPELL_SUMMON_MANA_FIEND_1   = 25681,
+    SPELL_SUMMON_MANA_FIEND_2   = 25682,
+    SPELL_SUMMON_MANA_FIEND_3   = 25683,
+    SPELL_DRAIN_MANA_EFFECT     = 25671,
+    SPELL_DRAIN_MANA_VISUAL     = 26639
 };
 
-enum Events
+enum MoamEvents
 {
-    EVENT_TRAMPLE           = 1,
-    EVENT_DRAIN_MANA        = 2,
-    EVENT_STONE_PHASE       = 3,
-    EVENT_STONE_PHASE_END   = 4,
-    EVENT_WIDE_SLASH        = 5,
+    EVENT_TRAMPLE               = 1,
+    EVENT_DRAIN_MANA,
+    EVENT_STONE_PHASE,
+    EVENT_ARCANE_ERUPTION
 };
 
-enum Actions
+enum MoamMisc
 {
-    ACTION_STONE_PHASE_START = 1,
-    ACTION_STONE_PHASE_END   = 2,
+    MAX_MANA_FIENDS             = 3
 };
 
-class boss_moam : public CreatureScript
+// 15340 - Moam
+struct boss_moam : public BossAI
 {
-    public:
-        boss_moam() : CreatureScript("boss_moam") { }
+    boss_moam(Creature* creature) : BossAI(creature, DATA_MOAM), _deadManaFiendCount(0) { }
+
+    void JustAppeared() override
+    {
+        DoCastSelf(SPELL_ZERO_MANA_FULL_HEALTH);
+    }
+
+    void Reset() override
+    {
+        _Reset();
+        _deadManaFiendCount = 0;
+    }
+
+    void JustEngagedWith(Unit* who) override
+    {
+        BossAI::JustEngagedWith(who);
+
+        Talk(EMOTE_AGGRO);
+
+        events.ScheduleEvent(EVENT_TRAMPLE, 10s, 25s);
+        events.ScheduleEvent(EVENT_DRAIN_MANA, 6s);
+        events.ScheduleEvent(EVENT_STONE_PHASE, 90s);
+        events.ScheduleEvent(EVENT_ARCANE_ERUPTION, 1s);
+    }
+
+    void SummonedCreatureDies(Creature* /*summon*/, Unit* /*killer*/) override
+    {
+        _deadManaFiendCount++;
 
-        struct boss_moamAI : public BossAI
+        if (_deadManaFiendCount == MAX_MANA_FIENDS)
         {
-            boss_moamAI(Creature* creature) : BossAI(creature, DATA_MOAM)
-            {
-                Initialize();
-            }
+            me->RemoveAurasDueToSpell(SPELL_ENERGIZE);
+            _deadManaFiendCount = 0;
+        }
+    }
 
-            void Initialize()
-            {
-                _isStonePhase = false;
-            }
+    void OnSpellCast(SpellInfo const* spell) override
+    {
+        if (spell->Id == SPELL_ARCANE_ERUPTION)
+        {
+            // Is there a spell for that? Maybe Zero Mana/Full Health but with a check to prevent setting full health if creature is in combat?
```

Comment:
i think that's a better handling than this one

Reply:
Without sniff I don't wanna handle it this way